sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/EqualsArgumentTypeCheck.kt

```diff
@@ -0,0 +1,85 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.lexer.KtTokens.THIS_KEYWORD
+import org.jetbrains.kotlin.psi.KtBinaryExpression
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtIsExpression
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.ANY_TYPE
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.api.EQUALS_METHOD_NAME
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.isAbstract
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val EQUALS_MATCHER = FunMatcher {
+    name = EQUALS_METHOD_NAME
+    withArguments(ANY_TYPE)
+}
+
+@Rule(key = "S2097")
+class EqualsArgumentTypeCheck : AbstractCheck() {
+
+    override fun visitClassBody(klass: KtClassBody, ctx: KotlinFileContext) {
+        val equalsMethod = klass.functions.find {
+            EQUALS_MATCHER.matches(it, ctx.bindingContext)
+        } ?: return
+
+        val parameter = equalsMethod.valueParameters[0]
+        val klassNames = (klass.parent as KtClass).collectDescendantsOfType<KtClass>().map {
+            it.name
+        }
+        val klassName = (klass.parent as KtClass).name
+
+        val hasIsExpression =
+            equalsMethod.collectDescendantsOfType<KtIsExpression> { parameter.name == (it.leftHandSide as KtNameReferenceExpression).getReferencedName() }
+                .flatMap { it.typeReference!!.text.split(".") } // class.childClass case
+                .any { it -> klassNames.contains(it) }
+
+
+        val binaryExpressions =
+            equalsMethod.collectDescendantsOfType<KtBinaryExpression> { it.operationToken == KtTokens.EQEQ || it.operationToken == KtTokens.EXCLEQ }
+
+        var isContained = false
+        run lit@{
+            binaryExpressions.forEach { binaryExpression ->
+                val left = binaryExpression.left!!.collectDescendantsOfType<KtNameReferenceExpression>().map { it.getReferencedName() }
+                val right = binaryExpression.right!!.collectDescendantsOfType<KtNameReferenceExpression>().map { it.getReferencedName() }
+
+                isContained =
+                    (left.contains(parameter.name) && (right.contains(klassName) || right.contains(THIS_KEYWORD.value))|| right.contains("javaClass"))
+                        ((left.contains(klassName) || left.contains(THIS_KEYWORD.value) || left.contains("javaClass")) && right.contains(parameter.name))
+                if (isContained)
+                    return@lit
+            }
+        }
+
+        equalsMethod.let {
+            if (!it.isAbstract() && !hasIsExpression && !isContained) ctx.reportIssue(it, "Add a type test to this method.")
```

Comment:
We should try to return from the rule eagerly when we see that a condition to report the issue is not met. E.g., we can check `.isAbstract` before doing most of the work in this rule and avoid doing any of it. Same for `hasIsExpression`, we actually know when this is true that we can abort the computation of `isContained`.